src/diffusers/loaders/single_file_utils.py

```diff
@@ -1249,369 +1192,256 @@ def create_text_encoder_from_open_clip_checkpoint(
-            num_in_channels = 7
+        image_size = kwargs.get("image_size", None)
+        image_size = set_image_size(cls, original_config, checkpoint, image_size=image_size, model_type=model_type)
 
-        else:
-            num_in_channels = 4
+        model_config = create_unet_diffusers_config(original_config, image_size=image_size)
 
-    image_size = set_image_size(
-        pipeline_class_name, original_config, checkpoint, image_size=image_size, model_type=model_type
-    )
-    unet_config = create_unet_diffusers_config(original_config, image_size=image_size)
-    unet_config["in_channels"] = num_in_channels
-    if upcast_attention is not None:
-        unet_config["upcast_attention"] = upcast_attention
+    elif config is not None:
+        model_config = config
 
-    diffusers_format_unet_checkpoint = convert_ldm_unet_checkpoint(checkpoint, unet_config, extract_ema=extract_ema)
-    ctx = init_empty_weights if is_accelerate_available() else nullcontext
+    else:
+        config = fetch_diffusers_config(checkpoint, subfolder="unet")
+        model_config = cls.load_config(**config)
 
+    if num_in_channels is not None:
+        deprecate("num_in_channels", "in_channels", "1.0.0")
+        model_config["in_channels"] = num_in_channels
+
+    ctx = init_empty_weights if is_accelerate_available() else nullcontext
     with ctx():
-        unet = UNet2DConditionModel(**unet_config)
+        model = cls.from_config(model_config, **kwargs)
 
+    diffusers_format_checkpoint = convert_ldm_unet_checkpoint(checkpoint, model_config, extract_ema=extract_ema)
     if is_accelerate_available():
-        from ..models.modeling_utils import load_model_dict_into_meta
-
-        unexpected_keys = load_model_dict_into_meta(unet, diffusers_format_unet_checkpoint, dtype=torch_dtype)
-        if unet._keys_to_ignore_on_load_unexpected is not None:
-            for pat in unet._keys_to_ignore_on_load_unexpected:
-                unexpected_keys = [k for k in unexpected_keys if re.search(pat, k) is None]
-
+        unexpected_keys = load_model_dict_into_meta(model, diffusers_format_checkpoint, dtype=torch_dtype)
         if len(unexpected_keys) > 0:
-            logger.warning(
-                f"Some weights of the model checkpoint were not used when initializing {unet.__name__}: \n {[', '.join(unexpected_keys)]}"
+            logger.warn(
+                f"Some weights of the model checkpoint were not used when initializing {cls.__name__}: \n {[', '.join(unexpected_keys)]}"
             )
+
     else:
-        unet.load_state_dict(diffusers_format_unet_checkpoint)
+        model.load_state_dict(diffusers_format_checkpoint)
 
     if torch_dtype is not None:
-        unet = unet.to(torch_dtype)
+        model.to(torch_dtype)
+
+    model.eval()
 
-    return {"unet": unet}
+    return model
 
 
-def create_diffusers_vae_model_from_ldm(
-    pipeline_class_name,
-    original_config,
+def create_diffusers_controlnet_from_ldm(
```

Comment:
If the high-level `create_diffusion_XXX_from_ldm` methods for model components share repetitive elements, should we try to unify them into a single function?